### PR TITLE
[MM-13833] Configured unit tests to log through t.Log

### PR DESCRIPTION
**Pull request by BK1603**
#### Summary
The logs from the app/server have been rerouted through `t.Log` and hence `debug` and `info` logs aren't printed until the verbose flag is specified making it easier to spot the test cases that failed.

#### Ticket Link
https://github.com/mattermost/mattermost-server/issues/10137

#### Checklist
- [x] Added or updated unit tests (required for all new features)

### CODE REVIEW

File: app/options.go [hunk 1]
```diff
@@ -69,6 +73,12 @@ func StartElasticsearch(s *Server) error {
 	return nil
 }
 
+func SetLogProxy(tb testing.TB) Option {
```
Comment (lieut-data) [inline app/options.go]: This method should probably be named `SetLogger` and accept just a `*mlog.Logger`. Then the calling code can leverage `NewLogProxy`, but the option becomes more generally useful for future code that might want to override the logging for other reasons.

File: mlog/log.go [hunk 2]
```diff
@@ -51,6 +52,20 @@ type Logger struct {
...
+type LogWriter struct {
+	tb testing.TB
+}
+
+func (lw *LogWriter) Write(b []byte) (int, error) {
+	n := len(b)
+	s := string(b)
+	lw.tb.Log(s)
+
+	// Assuming that the logs are written succesfuly
```
Comment (lieut-data) [inline mlog/log.go]: I think this is a perfectly fine assumption to make in this context, but a typo:
```suggestion
	// Assuming that the logs are written successfully
```

File: mlog/log.go [hunk 3]
```diff
@@ -51,6 +52,20 @@ type Logger struct {
 	fileLevel    zap.AtomicLevel
 }
 
+// LogWriter is an io.Writer that writes through t.Log
+type LogWriter struct {
+	tb testing.TB
+}
+
+func (lw *LogWriter) Write(b []byte) (int, error) {
+	n := len(b)
```
Comment (lieut-data) [inline mlog/log.go]: Suggest eliminating `n` and just inlining `len(b)`.

File: mlog/log.go [hunk 4]
```diff
@@ -51,6 +52,20 @@ type Logger struct {
...
+	s := string(b)
```
Comment (lieut-data) [inline mlog/log.go]: Suggest eliminating `s` and just inlining `string(b)`.

File: mlog/log.go [hunk 5]
```diff
@@ -109,6 +124,24 @@ func NewLogger(config *LoggerConfiguration) *Logger {
 	return logger
 }
 
+func NewLogProxy(tb testing.TB) *Logger {
```
Comment (lieut-data) [inline mlog/log.go]: Can you write a godoc for this function? Also, suggest calling this a `NewTestingLogger`.

File: mlog/log.go [hunk 6]
```diff
@@ -51,6 +52,20 @@ type Logger struct {
 	fileLevel    zap.AtomicLevel
 }
 
+// LogWriter is an io.Writer that writes through t.Log
```
Comment (lieut-data) [inline mlog/log.go]: Let's move this into a `testing.go` file in the same package, and rename as `TestingWriter`. Also, does the struct need to be exported or can it stay private?

**Comment by BK1603** [comment]
@lieut-data
Thanks a lot for looking into this! :smiley:

I'll fix everything that you said and push commits soon!

File: mlog/log.go (hunk 6 above)
Comment (BK1603) [reply mlog/log.go]: Should I move all the functions regarding the testing logger to the said file as well?

File: app/user_test.go [hunk 7]
```diff
@@ -18,11 +18,11 @@ import (
 
 	"github.com/mattermost/mattermost-server/einterfaces"
 	"github.com/mattermost/mattermost-server/model"
-	"github.com/mattermost/mattermost-server/model/gitlab"
+	oauthgitlab "github.com/mattermost/mattermost-server/model/gitlab"
```
[the remark this answers is not in the document]
Comment (BK1603) [reply app/user_test.go]: It was a merge conflict if i remember it correctly :sweat_smile:

File: mlog/log.go (hunk 5 above)
Comment (BK1603) [reply mlog/log.go]: Sure! :smile:

File: mlog/log.go (hunk 6 above)
Comment (BK1603) [reply mlog/log.go]: Moved everything related to `testingLogger` to a file named testing.go under the package mlog

File: mlog/testing.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+	"go.uber.org/zap/zapcore"
+)
+
+// testingWriter is an io.Writer that writes through t.Log
+type testingWriter struct {
+	tb testing.TB
+}
+
+func (tw *testingWriter) Write(b []byte) (int, error) {
+	tw.tb.Log(string(b))
```
Comment (lieut-data) [inline mlog/testing.go]: Let's trim off the newlines from the end of `b` here, as `t.Log` adds its own that clutters the output unnecessarily.

File: mlog/testing.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// testingWriter is an io.Writer that writes through t.Log
+type testingWriter struct {
+	tb testing.TB
+}
+
+func (tw *testingWriter) Write(b []byte) (int, error) {
```
Comment (lieut-data) [inline mlog/testing.go]: Testing this branch, it looks like `t.Log` prepends `testing.go: 20` in an attempt to identify the "source" of the log. This makes sense, as it's probably expecting `t.Log` to come from a test file whereas we're obviously latching onto that in a unique way. 

Adding `t.Helper` remove this frame from the stack, but then it just pops back to something deeper in `zap`. I don't think there's a solution to this, but just noting for completeness.

**Comment by lieut-data** [comment]
Also, @BK1603, to help me make the corresponding enterprise changes, I'm wondering if you can rename your branch from `master` to something like `testing_tlog`. We'd have to create a new PR, but then I /think/ I can match your branch name in our private repository to get everything synced up.

File: mlog/testing.go (hunk 9 above)
Comment (BK1603) [reply mlog/testing.go]: I'd like to look into this and see if there is a fix, I thought zap would pass up the caller too when it calls the write function, but now that you point it out, i think that's not how it works.

File: mlog/testing.go (hunk 8 above)
Comment (BK1603) [reply mlog/testing.go]: Sure! :smile:

**Comment by BK1603** [comment]
@lieut-data Removed the new line from `string(b)` in `testingLogger` and made the PR pass the build checks, I haven't change anything in `api4` or other packages as per your last edit, though if I need to just let me know :smile:

File: app/user_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (lieut-data) [reply app/user_test.go]: Oh, interesting, the package folder is named `gitlab` but it declares it's name as `oauthgitlab`. I suspect your editor fixed this for you. It's probably something we should fix at a higher level too, but no changes required here.

**Comment by jwilander** [comment]
@lieut-data is this ok to merge now or are there changes you want to make to the other repo/packages first?

**Comment by lieut-data** [comment]
No, this was ready to go as-is. There's some more work to do to extend this to other packages first, but that will also require enterprise changes. I'll likely work with @BK1603 to finish that off.

**Comment by BK1603** [comment]
Thanks a lot @lieut-data and @jwilander :smile:

**Pull request merged**